services/agents/registry.go

```diff
@@ -302,6 +313,58 @@ func authenticate(md *agentpb.AgentConnectMetadata, q *reform.Querier) (string,
 	return pointer.GetString(agent.RunsOnNodeID), nil
 }
 
+func addVMAgentToPMMAgent(q *reform.Querier, pmmAgentID, runsOnNodeID string) error {
+	// TODO remove it after fix
+	// https://jira.percona.com/browse/PMM-4420
+	if runsOnNodeID == "pmm-server" {
+		return nil
+	}
+	vmAgentType := models.VMAgentType
+	vmAgent, err := models.FindAgents(q, models.AgentFilters{PMMAgentID: pmmAgentID, AgentType: &vmAgentType})
+	if err != nil {
+		return status.Errorf(codes.Internal, "Can't get 'vmAgent' for pmm-agent with ID %q", pmmAgentID)
+	}
+	if len(vmAgent) == 0 {
+		if _, err := models.CreateAgent(q, models.VMAgentType, &models.CreateAgentParams{
+			PMMAgentID:  pmmAgentID,
+			PushMetrics: true,
+			NodeID:      runsOnNodeID,
+		}); err != nil {
+			return errors.Wrapf(err, "Can't create 'vmAgent' for pmm-agent with ID %q", pmmAgentID)
+		}
+	}
+	return nil
+}
+
+func removeVMAgentFromPMMAgent(q *reform.Querier, pmmAgentID string) error {
+	vmAgentType := models.VMAgentType
+	vmAgent, err := models.FindAgents(q, models.AgentFilters{PMMAgentID: pmmAgentID, AgentType: &vmAgentType})
```

Comment:
looks like it's better to do all these changes in one transaction.

Reply:
The same in `addVMAgentToPMMAgent`

Reply:
Agree. `Registry.register` should use `r.db.InTransaction` and call `authenticate` with `tx`, not `r.db.Querier`

Reply:
Thanks, added transaction.